File: lachesis/core/overlays/interprocedural.py
"""Context-specific argument, parameter, and return bindings."""
from __future__ import annotations

from collections import defaultdict

from ..composition import GraphDelta
from ..identities import stable_id
from ..query import GraphIndex


def _fact(evidence_ids: list[str], confidence: str = "exact") -> dict:
    return {
        "fact_origin": "core-inference",
        "confidence": confidence,
        "evidence_ids": list(dict.fromkeys(evidence_ids)),
    }


class InterproceduralContexts:
    """Keep each call/target binding distinct while stitching value flow."""

    overlay_id = "interprocedural-contexts"

    def applies(self, graph: dict) -> bool:
        return any(node.get("kind") == "call" for node in graph.get("nodes", []))

    def enrich(self, graph: dict) -> GraphDelta:
        index = GraphIndex(graph)
        nodes = []
        edges = []
        arguments_by_call: dict[str, list[dict]] = defaultdict(list)
        parameter_definitions: dict[str, list[str]] = defaultdict(list)
        returns_by_function: dict[str, list[str]] = defaultdict(list)
        bindings_by_call: dict[str, list[dict]] = defaultdict(list)

        for argument in index.nodes_of_kind("argument"):
            callsite_id = argument.get("properties", {}).get("callsite_id")
            if callsite_id:
                arguments_by_call[callsite_id].append(argument)
        for parameter in index.nodes_of_kind("parameter"):
            for definition in index.targets(parameter["id"], "DEFINES"):
                if definition.get("properties", {}).get("origin") == "parameter":
                    parameter_definitions[parameter["id"]].append(definition["id"])
        for edge in graph.get("edges", []):
            if edge.get("kind") == "RETURNS_VALUE":
                returns_by_function[edge["target"]].append(edge["source"])
            elif edge.get("kind") == "ARGUMENT_BINDS_PARAMETER":
                argument = index.nodes.get(edge["source"])
                callsite_id = argument and argument.get("properties", {}).get("callsite_id")
                if callsite_id:
                    bindings_by_call[callsite_id].append(edge)

        for call in index.nodes_of_kind("call", "construct"):
            call_id = call["id"]
            targets = list(dict.fromkeys(
                edge["target"]
                for edge in index.outgoing.get(call_id, [])
                if edge.get("kind") in {"INVOKES", "MAY_INVOKE"}
                and edge.get("target") in index.nodes
            ))
            contextual_targets = targets or [None]
            for target_id in contextual_targets:
                context_id = stable_id(
                    "core", self.overlay_id, "call-context",
                    call_id, target_id or "unresolved",
                )
                evidence = [call_id, *([target_id] if target_id else [])]
                fact = _fact(evidence, "exact" if target_id else "unresolved")
                nodes.append({
                    "id": context_id,
                    "kind": "call-context",
                    "label": f"context:{call.get('label', call_id)}",
                    "properties": {
                        **fact,
                        "callsite_id": call_id,
                        "caller_function_id": call.get("properties", {}).get("owner_function_id"),
                        "callee_function_id": target_id,
                        "resolution": call.get("properties", {}).get("resolution", "unresolved"),
                    },
                })
                edges.append({
                    "kind": "HAS_CALL_CONTEXT", "source": call_id,
                    "target": context_id, "properties": fact,
                })
                if target_id:
                    edges.append({
                        "kind": "CONTEXT_CALLS", "source": context_id,
                        "target": target_id, "properties": fact,
                    })

                for binding in bindings_by_call.get(call_id, []):
                    parameter_id = binding["target"]
                    parameter = index.nodes.get(parameter_id)
                    if not parameter:
                        continue
                    owner = parameter.get("properties", {}).get("owner_function_id")
                    if target_id and owner != target_id:
                        continue
                    argument_id = binding["source"]
                    position = binding.get("properties", {}).get("position")
                    binding_id = stable_id(
                        "core", self.overlay_id, "context-parameter",
                        context_id, argument_id, parameter_id,
                    )
                    binding_evidence = [call_id, argument_id, parameter_id]
                    binding_fact = _fact(binding_evidence)
                    nodes.append({
                        "id": binding_id,
                        "kind": "context-parameter",
                        "label": f"{call.get('label', 'call')} → {parameter.get('label', 'parameter')}",
                        "properties": {
                            **binding_fact,
                            "context_id": context_id,
                            "callsite_id": call_id,
                            "argument_id": argument_id,
                            "parameter_id": parameter_id,
                            "target_function_id": target_id,
                            "position": position,
                        },
                    })
                    edges.extend([
                        {
                            "kind": "BINDS_PARAMETER", "source": argument_id,
                            "target": binding_id,
                            "properties": {**binding_fact, "position": position},
                        },
                        {
                            "kind": "CONTEXTUALIZES", "source": binding_id,
                            "target": parameter_id, "properties": binding_fact,
                        },
                        {
                            "kind": "VALUE_FLOWS_TO", "source": argument_id,
                            "target": binding_id,
                            "properties": {**binding_fact, "reason": "context-argument"},
                        },
                    ])
                    for definition_id in parameter_definitions.get(parameter_id, []):
                        edges.append({
                            "kind": "VALUE_FLOWS_TO", "source": binding_id,
                            "target": definition_id,
                            "properties": {
                                **binding_fact,
                                "reason": "context-parameter",
                                "context_id": context_id,
                            },
                        })

                call_value_id = call.get("properties", {}).get("value_id")
                if not call_value_id or call_value_id not in index.nodes:
                    continue
                return_id = stable_id(
                    "core", self.overlay_id, "context-return", context_id, call_value_id,
                )
                return_evidence = [call_id, call_value_id, *returns_by_function.get(target_id, [])]
                return_fact = _fact(return_evidence, "exact" if target_id else "unresolved")
                nodes.append({
                    "id": return_id,
                    "kind": "context-return",
                    "label": f"return:{call.get('label', call_id)}",
                    "properties": {
                        **return_fact,
                        "context_id": context_id,
                        "callsite_id": call_id,
                        "callee_function_id": target_id,
                        "call_value_id": call_value_id,
                    },
                })
                edges.append({
                    "kind": "CONTEXT_RETURNS", "source": context_id,
                    "target": return_id, "properties": return_fact,
                })
                for source_return_id in returns_by_function.get(target_id, []):
                    edges.append({
                        "kind": "VALUE_FLOWS_TO", "source": source_return_id,
                        "target": return_id,
                        "properties": {
                            **return_fact,
                            "reason": "context-return",
                            "context_id": context_id,
                        },
                    })
                edges.append({
                    "kind": "VALUE_FLOWS_TO", "source": return_id,
                    "target": call_value_id,
                    "properties": {
                        **return_fact,
                        "reason": "context-call-result",
                        "context_id": context_id,
                    },
                })

        return GraphDelta(self.overlay_id, nodes, edges)

File: native/lifetime_kernel/src/interprocedural.rs
//! Native context-specific argument/parameter/return bindings.

use hashbrown::{HashMap, HashSet};
use crate::graph_proto;
use crate::pass2::{self, Delta, Graph};

fn text<'a>(node: &'a pass2::Node, key: &str) -> Option<&'a str> {
    node.properties.iter().find_map(|field| {
        if field.key != key { return None; }
        match field.value.as_ref()?.kind.as_ref()? {
            graph_proto::value::Kind::Text(value) => Some(value.as_str()), _ => None,
        }
    })
}
fn integer(node: &pass2::Node, key: &str) -> Option<i64> {
    node.properties.iter().find_map(|field| {
        if field.key != key { return None; }
        match field.value.as_ref()?.kind.as_ref()? {
            graph_proto::value::Kind::Integer(value) => Some(*value), _ => None,
        }
    })
}
fn edge_text<'a>(edge: &'a pass2::Edge, key: &str) -> Option<&'a str> {
    edge.properties.iter().find_map(|field| {
        if field.key != key { return None; }
        match field.value.as_ref()?.kind.as_ref()? {
            graph_proto::value::Kind::Text(value) => Some(value.as_str()), _ => None,
        }
    })
}
fn edge_integer(edge: &pass2::Edge, key: &str) -> Option<i64> {
    edge.properties.iter().find_map(|field| {
        if field.key != key { return None; }
        match field.value.as_ref()?.kind.as_ref()? {
            graph_proto::value::Kind::Integer(value) => Some(*value), _ => None,
        }
    })
}
fn list_field(key: &str, values: &[String]) -> graph_proto::Field {
    graph_proto::Field { key: key.to_owned(), value: Some(graph_proto::Value { kind: Some(
        graph_proto::value::Kind::List(graph_proto::ListValue { values: values.iter().map(|v| graph_proto::Value {
            kind: Some(graph_proto::value::Kind::Text(v.clone())) }).collect() })
    ) }) }
}
fn fact(evidence: &[String], confidence: &str) -> Vec<graph_proto::Field> {
    vec![pass2::text_field("fact_origin", "core-inference"), pass2::text_field("confidence", confidence), list_field("evidence_ids", evidence)]
}
fn node(id: String, kind: &str, label: String, properties: Vec<graph_proto::Field>) -> graph_proto::NodeRecord {
    graph_proto::NodeRecord { id, kind: kind.to_owned(), label, properties, tier: String::new() }
}
fn edge(kind: &str, source: &str, target: &str, properties: Vec<graph_proto::Field>) -> graph_proto::EdgeRecord {
    graph_proto::EdgeRecord { kind: kind.to_owned(), source: source.to_owned(), target: target.to_owned(), properties, source_tier: String::new(), relationship_class: String::new() }
}

pub(crate) fn enrich(graph: &Graph) -> Delta {
    let mut arguments_by_call: HashMap<u32, Vec<u32>> = HashMap::new();
    let mut parameter_defs: HashMap<u32, Vec<u32>> = HashMap::new();
    let mut returns_by_function: HashMap<u32, Vec<u32>> = HashMap::new();
    let mut bindings_by_call: HashMap<u32, Vec<usize>> = HashMap::new();
    for node in &graph.nodes {
        if graph.kind(node.kind) == "argument" {
            if let Some(call) = text(node, "callsite_id").and_then(|id| graph.symbol(id)) { arguments_by_call.entry(call).or_default().push(node.id); }
        }
    }
    for (edge_index, item) in graph.edges.iter().enumerate() {
        match graph.edge_kind(item).as_str() {
            "RETURNS_VALUE" => { returns_by_function.entry(item.target).or_default().push(item.source); }
            "ARGUMENT_BINDS_PARAMETER" => {
                if let Some(argument) = graph.node_by_id.get(&item.source).map(|index| &graph.nodes[*index]) {
                    if let Some(call) = text(argument, "callsite_id").and_then(|id| graph.symbol(id)) {
                        bindings_by_call.entry(call).or_default().push(edge_index);
                    }
                }
            }
            "DEFINES" => {
                let target = graph.node_by_id.get(&item.target).map(|index| &graph.nodes[*index]);
                if target.is_some_and(|node| text(node, "origin") == Some("parameter")) { parameter_defs.entry(item.source).or_default().push(item.target); }
            }
            _ => {}
        }
    }
    let mut nodes = Vec::new(); let mut edges = Vec::new(); let mut emitted = HashSet::new();
    for call in &graph.nodes {
        if !matches!(graph.kind(call.kind), "call" | "construct") { continue; }
        let targets: Vec<u32> = graph.outgoing[graph.node_by_id[&call.id]].iter().filter_map(|edge_index| {
            let item = &graph.edges[*edge_index];
            (matches!(graph.edge_kind(item).as_str(), "INVOKES" | "MAY_INVOKE") && graph.node_by_id.contains_key(&item.target)).then_some(item.target)
        }).collect::<HashSet<_>>().into_iter().collect();
        let contextual_targets = if targets.is_empty() { vec![None] } else { targets.into_iter().map(Some).collect() };
        let call_id = graph.id(call.id).to_owned();
        for target in contextual_targets {
            let target_id = target.map(|id| graph.id(id).to_owned());
            let context_id = pass2::stable_id("core", "interprocedural-contexts", "call-context", &[&call_id, target_id.as_deref().unwrap_or("unresolved")]);
            if !emitted.insert(context_id.clone()) { continue; }
            let confidence = if target_id.is_some() { "exact" } else { "unresolved" };
            let mut evidence = vec![call_id.clone()]; if let Some(id) = &target_id { evidence.push(id.clone()); }
            let context_fact = fact(&evidence, confidence);
            let label = format!("context:{}", call.label);
            nodes.push(node(context_id.clone(), "call-context", label, [context_fact.clone(), vec![pass2::text_field("callsite_id", &call_id)], target_id.as_ref().map(|id| vec![pass2::text_field("callee_function_id", id)]).unwrap_or_default()].concat()));
            edges.push(edge("HAS_CALL_CONTEXT", &call_id, &context_id, context_fact.clone()));
            if let Some(target_id) = &target_id { edges.push(edge("CONTEXT_CALLS", &context_id, target_id, context_fact.clone())); }
            for edge_index in bindings_by_call.get(&call.id).into_iter().flatten() {
                let binding = &graph.edges[*edge_index];
                let parameter_id = binding.target;
                let Some(parameter) = graph.node_by_id.get(&parameter_id).map(|index| &graph.nodes[*index]) else { continue };
                if let Some(target) = target { if text(parameter, "owner_function_id").and_then(|id| graph.symbol(id)) != Some(target) { continue; } }
                let argument_id = binding.source;
                let argument_text = graph.id(argument_id).to_owned(); let parameter_text = graph.id(parameter_id).to_owned();
                let binding_id = pass2::stable_id("core", "interprocedural-contexts", "context-parameter", &[&context_id, &argument_text, &parameter_text]);
                let binding_evidence = vec![call_id.clone(), argument_text.clone(), parameter_text.clone()];
                let binding_fact = fact(&binding_evidence, "exact");
                let mut properties = [binding_fact.clone(), vec![pass2::text_field("context_id", &context_id), pass2::text_field("callsite_id", &call_id), pass2::text_field("argument_id", &argument_text), pass2::text_field("parameter_id", &parameter_text)]].concat();
                if let Some(position) = edge_integer(binding, "position") { properties.push(pass2::integer_field("position", position)); }
                nodes.push(node(binding_id.clone(), "context-parameter", format!("{} → {}", call.label, parameter.label), properties));
                edges.push(edge("BINDS_PARAMETER", &argument_text, &binding_id, binding_fact.clone()));
                edges.push(edge("CONTEXTUALIZES", &binding_id, &parameter_text, binding_fact.clone()));
                edges.push(edge("VALUE_FLOWS_TO", &argument_text, &binding_id, [binding_fact.clone(), vec![pass2::text_field("reason", "context-argument")]].concat()));
                for definition in parameter_defs.get(&parameter_id).into_iter().flatten() {
                    let definition_text = graph.id(*definition).to_owned();
                    edges.push(edge("VALUE_FLOWS_TO", &binding_id, &definition_text, [binding_fact.clone(), vec![pass2::text_field("reason", "context-parameter"), pass2::text_field("context_id", &context_id)]].concat()));
                }
            }
            let Some(call_value) = text(call, "value_id").and_then(|id| graph.symbol(id)) else { continue };
            let call_value_text = graph.id(call_value).to_owned();
            let return_id = pass2::stable_id("core", "interprocedural-contexts", "context-return", &[&context_id, &call_value_text]);
            let mut return_evidence = vec![call_id.clone(), call_value_text.clone()];
            return_evidence.extend(target.and_then(|id| returns_by_function.get(&id)).into_iter().flatten().map(|id| graph.id(*id).to_owned()));
            let return_fact = fact(&return_evidence, confidence);
            nodes.push(node(return_id.clone(), "context-return", format!("return:{}", call.label), [return_fact.clone(), vec![pass2::text_field("context_id", &context_id), pass2::text_field("callsite_id", &call_id), pass2::text_field("call_value_id", &call_value_text)]].concat()));
            edges.push(edge("CONTEXT_RETURNS", &context_id, &return_id, return_fact.clone()));
            if let Some(target) = target { for source in returns_by_function.get(&target).into_iter().flatten() { let source_text = graph.id(*source).to_owned(); edges.push(edge("VALUE_FLOWS_TO", &source_text, &return_id, [return_fact.clone(), vec![pass2::text_field("reason", "context-return"), pass2::text_field("context_id", &context_id)]].concat())); } }
            edges.push(edge("VALUE_FLOWS_TO", &return_id, &call_value_text, [return_fact, vec![pass2::text_field("reason", "context-call-result"), pass2::text_field("context_id", &context_id)]].concat()));
        }
    }
    Delta { nodes, edges }
}
